### DISCUSSION TITLE
Unpack a Generic with a TypedDict bound

### DISCUSSION THREAD

**Discussion by joknoll** [microsoft/pyright#7317] [Q&A] [answered]
In Python 3.12, the introduction of the Unpack type hint and its ability to unpack TypedDicts has opened up new possibilities for type-safe and expressive code. However, when combined with generics, some intriguing questions and potential challenges arise.

```python
from typing import TypedDict, Unpack

@dataclass
class Message[T: TypedDict]:
    header: int
    data: T

    def serialize(self, **data: Unpack[T]) -> bytes:
        self.data = data
        return b"..."
```
This code defines a generic message that has its generic `T` bound to `TypedDict`, which is necessary for the `Unpack`.
The `serialize` method tries to unpack the data kwargs using Unpack[T], expecting them to match the fields of the specific TypedDict.

pyright gives following message `error: Expected TypedDict type argument for Unpack (reportGeneralTypeIssues)`, which is a bit confusing because T is bound to TypedDict.

It seems like the current implementation of Unpack doesn't interact with generics. I'm curious if there's a specific reason for this or if there are plans to support generics in the future? The Unpack [PEP 692](https://peps.python.org/pep-0692/) doesn't address the use of generics, but I think it could enhance its flexibility and make it applicable to even more use cases.

**Comment by erictraut** [comment, marked as answer]
There are two problems here. First, `TypedDict` cannot be used as an upper bound for a TypeVar. `TypedDict` is a special form that's used to define a structural type. It's not an actual class that participates in a class hierarchy. There are no classes that are subtypes of `TypedDict`.

Second, you cannot use a generic in a `**kwargs: Unpack[T]` construct. [PEP 692](https://peps.python.org/pep-0692/) is clear that the type `T` must be an actual TypedDict class, not a type variable.

**Reply by erictraut** [reply]
There is a small bug here. Pyright should emit an error if you attempt to use `TypedDict` as the bound for a TypeVar. It correctly generates this error if you use the traditional `TypeVar` constructor, but it doesn't currently do it for the new-style (PEP 695) type parameter syntax. I'll fix this so it properly emits an error in both cases.

**Reply by joknoll** [reply]
Thanks for the fast reply! Good you have found a bug as a side effect.
I understand the points you made, a solution would be to just make the function signature `serialize(self, data: T) -> bytes` without `**` and `Unpack`.
The kwargs would have been handy, but don’t seem to be the right tool here.
